docs/decisions/0002-feature-based-application-organization.rst

```diff
@@ -0,0 +1,157 @@
+
+Decision
+--------
+
+**Following the spirit of these principles is more important than following them to the letter.**
+
+These rules are guidelines. It won't always be reasonable or necessary to follow them to the letter. They provide a set of tools for dealing with complexity. It follows, then, that if you don't have complexity, then you may not need the tools.
+
+Primary guiding principles
+==========================
+
+**1. Code is organized into feature directories.**
+
+A feature is a logical or semantically significant grouping of code; what comprises a feature is subjective.  It's also emergent - if code tends to be related or change together, then it's probably part of the same feature.
+
+It's unlikely to be worth agonizing over your feature breakdown; time will tell what's correct moreso than overthinking it. That said, a sufficiently complex set of features will need a similarly robust taxonomy and organizational hierarchy. (This document endeavors to help inform that hierarchy.)  A nice rule of thumb is that a feature should conceptually be able to be extracted into its own npm package with minimal effort.
+
+**2. Create strict module boundaries**
+
+A feature should have a public interface exposed via an index.js file in the feature directory.  Consumers of a feature should limit themselves to importing only from the public exports.
+
+Modules are configured from above. For react components, this is giving them props, for services, this is giving it apiClient/configuration, etc.
+
+**3. Avoid circular dependencies**
+
+Circular dependencies are unresolvable in webpack, and will result in something being imported as 'undefined'.  They're also incredibly difficult and frustrating to track down.  Properly factoring your features and supporting modules should help avoid these sorts of issues.  In general, a feature should never need to import from its parent, or from its lineage (grandparents, great-grandparents, etc.).  Furthermore, think about how "general purpose" a given feature or module is.  If you find yourself importing from a domain-specific feature in your general utility module, then something is probably ill-factored.
+
+Secondary principles
+====================
+
+**A. Separate business logic from components**
+
+In order to isolate our view layer (react) from business logic inherent in our global state, APIs, and side effect management, we want to adopt the "ducks" organization (see references).  This involves isolating business logic into a
+sub directory of a feature.  See "B" below for an example.
+
+**B. Files in a feature directory are named by type of code**
+
+Since directory names describe features, file names describe types of code.
+
+::
+
+  /profile
+    /index.js // public interface
+    /ProfilePage.jsx // view
+    /ProfilePhotoUploader.jsx // supporting view
+    /ducks // business logic
```

Comment:
Interesting that we're calling the "ducks" folder as "business logic".  In my mind, I considered it to be a directory of boiler-plate plumbing logic - of technical concerns with selectors, reducers, etc.

The term "business logic", as used in DDD and Clean Architecture, generally refers to domain logic and code that is specific to your feature/business domain/etc.

Thoughts?

Reply:
As an aside, I strongly dislike using the term "ducks" at all.

That said, I wasn't 100% sold on calling it "business logic", as that didn't seem quite right.  Frankly I thought I'd heard you call it that in conversation so I think I got the term in my head and couldn't figure out what else to call it. ;)

I think I'd like to replace "ducks" with the word "data", as a more generic term that mostly captures what's in there.  Data plumbing logic. 👍